src/main/java/com/limechain/trie/structure/TrieStructure.java

```diff
@@ -1,30 +1,483 @@
+        }
+
+        Collections.reverse(path);
+
+        return path;
+
+//        // Alternatively (and more functionally :D):
+//        return Stream.iterate(this.getNodeAtIndexInner(targetNodeIndex).parentIndex,
+//                Objects::nonNull,
+//                index -> this.getNodeAtIndexInner(index.getValue0()).parentIndex)
+//            .map(Pair::getValue0)
+//            .collect(Collectors.collectingAndThen(Collectors.toList(), list -> {
+//                Collections.reverse(list);
+//                return list;
+//            }));
+    }
+
+    /**
+     * Returns true if the structure of this trie is the same as the structure of {@code other}.
+     * Everything is compared for equality except for the {@link TrieNode#userData}s.
+     *
+     * @implNote This method first compares the sizes of the two trie structures,
+     *           and if they don't match, it early returns false.
+     *           Otherwise, it performs the expensive operation of traversing both tries
+     *           in order to identify a potential mismatch.
+     */
+    public <U> boolean structurallyEquals(TrieStructure<U> other) {
+        if (this.nodes.size() != other.nodes.size()) {
+            return false;
+        }
+
+        var thisIter = this.iteratorOrdered();
+        var otherIter = other.iteratorOrdered();
+
+        while (true) {
+            if (!thisIter.hasNext() && !otherIter.hasNext()) {
+                return true;
+            }
+
+            var thisNode = this.getNodeAtIndexInner(thisIter.next().getValue());
+            var otherNode = other.getNodeAtIndexInner(otherIter.next().getValue());
+
+            if (thisNode.hasStorageValue != otherNode.hasStorageValue) {
+                return false;
+            }
+
+            // Check if the parents match
+            {
+                var thisNodeParent = thisNode.parent;
+                var otherNodeParent = otherNode.parent;
+
+                //NOTE: this is awful, so a quick explainer:
+                //  We want to return false in all cases except:
+                //    - both parents are null;
+                //    - both parents are not null and the two nodes' child indices within them are the same
+                //  Reference: https://github.com/smol-dot/smoldot/blob/200214a571af30b5fa3997aea988451adc235ed0/lib/src/trie/trie_structure.rs#L615
+                if ((thisNodeParent != null || otherNodeParent != null)
+                    && (thisNodeParent == null
+                        || otherNodeParent == null
+                        || !thisNodeParent.childIndexWithinParent().equals(otherNodeParent.childIndexWithinParent()))) {
+                    return false;
+                }
+            }
+
+            if (!thisNode.partialKey.equals(otherNode.partialKey)) {
+                return false;
+            }
+        }
+    }
+
+    // TODO: Figure out how to sensibly deprecate this or whether it should be equivalent to structurallyEquals
```

Comment:
This overridden deprecated `equals` to be removed, but I'm not sure how. I suspect my initial idea to dissuade the users to ever use it is rather stupid.

The idea is: we want to have two separate "equals" methods - one is `structurallyEquals`, the other doesn't exist yet, but will include comparison of the `userData` as well. 

Figure out a decent decision.

Reply:
Can't we override the equals method to have the body of `structurallyEquals` (or the second one since it's checking for more fields) for example?

Reply:
> Can't we override the equals method to have the body of `structurallyEquals` (or the second one since it's checking for more fields) for example?

Of course we can... but do we want? 😅 I'm personally against the existence of a method called `equals()` in a scenario where no singular sensible equality is possible, I'd prefer different names for explicitness. But Java will always require an `equals()`, so it's our call. I'm leaving this up to you - easiest solution would be to implement it as `structurallyEquals`. Do we proceed that way?

Reply:
I'm not sure what implementing `equals` in either way would imply for the advisably corresponding `hashCode`. I feel like currently no sensible `hashCode` implementation could exist (without going in a bit more depth for unused stuff right now), so overriding `equals` with different behaviour, but not `hashChode` would break the contract on the notion of 'equality'. This makes me think we should leave `equals` deprecated.

Reply:
I see three options:
1) throw `UnsupportedOperationException` from both `equals` and `hashCode`;
2) just leave them as the default behaviour (i.e. don't override);
3) actually figure out what would make sense for `equals` and `hashCode`. Let's say equals is the easier part. But this is work I'm willing to put off for later.

I'm going with option 2.

Reply:
Regarding option 3 (the actual good approach):
we could make `equals` a superset of `structurallyEquals`, but including comparison of `userData`s. Easy way to achieve this:
1) implement `equals` for `TrieNode`'s, basically similar to a record's `equals`, comparing fields;
2) within the TrieStructure, iterate nodes lexicographically and compare the collections obtained.

But this'll remain for later. I'm intentionally not resolving this conversation.